storehaus-cassandra/src/main/scala/com/twitter/storehaus/cassandra/cql/CASStore.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2014 Twitter, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.twitter.storehaus.cassandra.cql
+
+import com.twitter.util.{Future, Return, Throw}
+
+/**
+ * represents store supporting Compare/Check and Set. This allows to implement 
+ * distributed asynchronous "transaction-like" operations. E.g. a distributed 
+ * lock implemented with this and Futures is a snap.
+ */
+trait CASStore[T, K, V] {
+  
+  def get(k: K)(implicit ev1: Equiv[T]): Future[Option[(V, T)]]
+  
+  def cas(token: Option[T], kv: (K, V))(implicit ev1: Equiv[T]): Future[Boolean]  
```

Comment:
The semantics of CASStore was indeed already subject to an internal emotional debate a few month ago. Added a comment and some tests showing the semantics which is implemented: None means, that there must not exist a token previously (and the operation is only performed, if that is the case) and the store sets the first one using some token-factory; while providing Some(token) means that the store needs to check for that token and a new one is being set during the operation iff the previously existing token is stored before/during the start of the operation. One could - of course - define a variety of other semantics.